### PR TITLE
Scan plugins

### PR DESCRIPTION
**Pull request by Meakk**

### CODE REVIEW

File: application/F3DSystemTools.cxx [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+namespace F3DSystemTools
+{
+//----------------------------------------------------------------------------
+fs::path GetApplicationDirectory()
+{
+#if defined(_WIN32)
+  wchar_t wc[1024] = { 0 };
+  GetModuleFileNameW(NULL, wc, 1024);
+  return fs::path(wc);
```
Comment (mwestphal) [inline application/F3DSystemTools.cxx:28]: maybe something like ?

```
        std::vector<wchar_t> wc(1024);
        if (GetModuleFileNameW(
          GetModuleHandle(nullptr), wc.data(), static_cast<DWORD>(wc.size())))
        {
          return fs::path(wc);
        }
```

File: application/F3DSystemTools.cxx [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+#include "log.h"
+
+#include <cstring>
+#include <filesystem>
+
+#if defined(_WIN32)
+// clang-format off
+#include <windows.h>
+#include <libloaderapi.h>
```
Comment (mwestphal) [inline application/F3DSystemTools.cxx:11]: is this actually needed ?

File: application/F3DSystemTools.cxx [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+fs::path GetApplicationDirectory()
+{
+#if defined(_WIN32)
+  wchar_t wc[1024] = { 0 };
+  GetModuleFileNameW(NULL, wc, 1024);
+  return fs::path(wc);
+#else
+#ifdef __APPLE__
+  uint32_t size = 1024;
+  char buffer[size];
```
Comment (mwestphal) [inline application/F3DSystemTools.cxx:32]: std::vector<char> ?

File: application/F3DSystemTools.cxx [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+#ifdef __APPLE__
+#include <mach-o/dyld.h>
+#endif
+
+namespace fs = std::filesystem;
+
+namespace F3DSystemTools
+{
+//----------------------------------------------------------------------------
+fs::path GetApplicationDirectory()
```
Comment (mwestphal) [inline application/F3DSystemTools.cxx:23]: GetApplicationPath ?

File: application/F3DSystemTools.cxx [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+  uint32_t size = 1024;
+  char buffer[size];
+  if (_NSGetExecutablePath(buffer, &size) != 0)
+  {
+    f3d::log::error("Executable is too long to recover path to configuration file");
+    return fs::path();
+  }
+  return fs::path(buffer);
+#else
+  return fs::canonical("/proc/self/exe");
```
Comment (mwestphal) [inline application/F3DSystemTools.cxx:40]: i think you need to try catch the whole block ?

File: application/testing/CMakeLists.txt [hunk 6]
```diff
@@ -620,3 +620,8 @@ f3d_test(NAME TestBadRefNoOutput DATA cow.vtp ARGS --ref ${CMAKE_SOURCE_DIR}/tes
 
 # Test failure with a bad interaction play file, please do not create a dummy.log
 f3d_test(NAME TestPlayNoFile DATA cow.vtp ARGS --interaction-test-play=${CMAKE_BINARY_DIR}/Testing/Temporary/dummy.log WILL_FAIL)
+
+# Test scan plugins
+if(NOT F3D_MACOS_BUNDLE)
+  f3d_test(NAME TestScanPlugins ARGS --scan-plugins NO_RENDER NO_BASELINE REGEXP " - native")
+endif()
```
Comment (mwestphal) [inline application/testing/CMakeLists.txt:627]: I think other tests do not make sense in MACOS bundle then ?

File: application/testing/CMakeLists.txt [hunk 7]
```diff
@@ -620,3 +620,8 @@ f3d_test(NAME TestBadRefNoOutput DATA cow.vtp ARGS --ref ${CMAKE_SOURCE_DIR}/tes
 
 # Test failure with a bad interaction play file, please do not create a dummy.log
 f3d_test(NAME TestPlayNoFile DATA cow.vtp ARGS --interaction-test-play=${CMAKE_BINARY_DIR}/Testing/Temporary/dummy.log WILL_FAIL)
+
+# Test scan plugins
+if(NOT F3D_MACOS_BUNDLE)
+  f3d_test(NAME TestScanPlugins ARGS --scan-plugins NO_RENDER NO_BASELINE REGEXP " - native")
```
Comment (mwestphal) [inline application/testing/CMakeLists.txt:626]: test with an actual dynamic plugin like exodus ?

File: library/src/engine.cxx [hunk 8]
```diff
@@ -247,6 +250,46 @@ void engine::autoloadPlugins()
...
+  if (dir.Load(pluginPath))
+  {
+    for (unsigned long i = 0; i < dir.GetNumberOfFiles(); i++)
+    {
+      std::string currentFile = dir.GetFile(i);
+      if (std::equal(ext.rbegin(), ext.rend(), currentFile.rbegin()))
+      {
+        std::string fullPath = dir.GetPath();
+        fullPath += "/";
+        fullPath += currentFile;
```
Comment (mwestphal) [inline library/src/engine.cxx:269]: use filesystem

File: library/src/engine.cxx [hunk 9]
```diff
@@ -247,6 +250,46 @@ void engine::autoloadPlugins()
...
+          auto name = root.find("name");
+
+          if (name != root.end() && name.value().is_string())
+          {
+            pluginNames.push_back(name.value().get<std::string>());
+          }
+        }
+        catch (...)
+        {
+          // silently skip current file
```
Comment (mwestphal) [inline library/src/engine.cxx:284]: log an error ?

File: python/testing/CMakeLists.txt [hunk 10]
```diff
@@ -13,12 +19,13 @@ endif()
 
 list(APPEND pyf3dTestsNoRender_list
      TestPythonOptions
+     TestPythonPlugins
```
Comment (mwestphal) [inline python/testing/CMakeLists.txt:22]: wont that fails on macos ?

**Review by mwestphal — CHANGES_REQUESTED** [review]

**Comment by mwestphal** [comment]
I dont see any CMake code for MACOS bundle, I can take care of it though

File: application/F3DSystemTools.cxx (hunk 1 above)
Comment (Meakk) [reply application/F3DSystemTools.cxx:28]: I haven't changed this code, it's just moved. I can use a `std::array` though if you prefer.

Comment (mwestphal) [reply application/F3DSystemTools.cxx:28]: I'd just like it to be coherent as it is present in engine.cxx as well. I will take care of it in another PR though

File: library/src/engine.cxx (hunk 8 above)
Comment (Meakk) [reply library/src/engine.cxx:269]: There is no usage of filesystem in libf3d yet. We will do the migration later.

File: application/testing/CMakeLists.txt (hunk 6 above)
Comment (mwestphal) [reply application/testing/CMakeLists.txt:627]: TestPluginName and TestPluginFullPath ?

File: library/src/engine.cxx (hunk 8 above)
Comment (mwestphal) [reply library/src/engine.cxx:269]: indeed

**Comment by Meakk** [comment]
> I dont see any CMake code for MACOS bundle, I can take care of it though

What do you mean?

File: application/testing/CMakeLists.txt (hunk 6 above)
Comment (Meakk) [reply application/testing/CMakeLists.txt:631]: No these ones are fine. The problem with the bundle is only related to the plugin json files.

**Review by mwestphal — APPROVED** [review]

**Comment by Meakk** [comment]
Right. Let's do it later in another PR.

File: application/F3DSystemTools.cxx [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+  }
+  f3d::log::error("Cannot retrieve application path");
+  return fs::path();
+#else
+#ifdef __APPLE__
+  uint32_t size = 1024;
+  std::array<char, 1024> buffer;
+  if (_NSGetExecutablePath(buffer.data(), &size) != 0)
+  {
+    f3d::log::error("Executable is too long to recover path to configuration file");
```
Comment (mwestphal) [inline application/F3DSystemTools.cxx:37]: wrong log

Comment (mwestphal) [reply application/F3DSystemTools.cxx:37]: f3d::log::error("Cannot retrieve application path");

**Pull request merged**